### PR TITLE
Remove jQuery from browse

### PR DESCRIPTION
**Pull request by andysellick**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Remove jQuery from the browse columns Javascript at https://www.gov.uk/browse

The original code relied heavily on jQuery's `animate` function and a lot of promise resolving after ajax calls to ensure that column changes and animations happened at the correct time. This was quite complex. There is no native JS equivalent to `animate` and writing one just for this page would be overkill. Instead, the code has been simplified and uses XMLHttpRequest and CSS transitions to reproduce the behaviour. This is a significant change but should result in no difference to the user.

Also adds a few more tests, as there weren't many. There still probably aren't enough 😢 

There's a lot of code changes involved. I'd recommend glancing through it commit by commit to understand how it has changed. Things to test when reviewing:

- visually matches existing browse
- browser back button works as expected (also forward button)
- mobile view should only show the currently open 'pane'
- reloading the page after changing the state (e.g. browse through the page, refresh, check still looks the same)
- resizing the browser window and reloading after resizing, then resizing again
- check that both curated views (normal) and A-Z views (e.g. births deaths marriages) appear correctly
- analytics events still fire correctly and contain the right data

## Why

We're taking jQuery out of GOV.UK and this is the last script in `collections` that uses it.

## Visual changes

Should be no changes, apart from the spacing increase beneath headings for issue 2597.

## Related issues

- Fixes https://github.com/alphagov/collections/issues/2597
- Fixes https://github.com/alphagov/collections/issues/1483
- Fixes https://github.com/alphagov/collections/issues/2414

Trello card: https://trello.com/c/mFgNxxIB/487-remove-jquery-from-collections-browse-columnsjs

### CODE REVIEW

File: app/assets/javascripts/browse-columns.js [hunk 1]
```diff
@@ -1,395 +1,323 @@
...
+    } else {
+      return $link.textContent
+    }
+  }
+
+  BrowseColumns.prototype.setTitle = function (title) {
+    document.title = title + ' - GOV.UK'
+  }
+
+  BrowseColumns.prototype.setContentIDMetaTag = function (contentId) {
```
Comment (kevindew) [inline app/assets/javascripts/browse-columns.js:227]: It'd be more idiomatic to our regular naming conventions to call this `setContentIdMetaTag`

**Review by kevindew — COMMENTED** [review]
Amazing work Andy 👍  Lots of great improvements and modernisations here. Well done.

I'll be honest, I've found this difficult to review as so much is changing in one PR with the original code being rather terse, so I ended up jumping between this branch and the existing one to try compare - so it's a bit of a cursory review. I added a few minor comments.

Trying it out locally, it all seems to work well but the animations I was expecting didn't seem to be working:

Dec-01-2021 09-33-30

File: app/assets/javascripts/browse-columns.js [hunk 2]
```diff
@@ -1,395 +1,323 @@
...
-      var state = e.originalEvent.state
-      var loadPromise
-      if (!state) { // state will be null if there was no state set
-        state = this.parsePathname(window.location.pathname)
-      }
 
-      if (this.lastState.slug === state.slug) {
-        return // nothing has changed
-      }
+  BrowseColumns.prototype.navigate = function (e) {
```
[the remark this answers is not in the document]
Comment (andysellick) [reply app/assets/javascripts/browse-columns.js]: Good point, changed.

File: app/assets/javascripts/browse-columns.js [hunk 3]
```diff
@@ -1,395 +1,323 @@
...
+    var target = e.target
+    var clicked = target.tagName === 'A' ? target : target.closest('a')
```
[the remark this answers is not in the document]
Comment (andysellick) [reply app/assets/javascripts/browse-columns.js:51]: Good spot. I've now added the polyfill for `closest`.

File: app/assets/javascripts/browse-columns.js (hunk 1 above)
Comment (andysellick) [reply app/assets/javascripts/browse-columns.js]: Changed, thanks.

File: app/assets/javascripts/browse-columns.js [hunk 4]
```diff
@@ -1,395 +1,323 @@
...
+      this.$module.insertBefore(this.$section, this.$module.firstChild)
+    }
 
-    $(window).on('popstate', this.popState.bind(this))
+    if (!this.$subsection) {
+      this.$subsection = document.createElement('div')
+      this.$subsection.setAttribute('id', 'subsection')
+      this.$subsection.setAttribute('class', 'browse__subsection-pane')
+      this.$module.insertBefore(this.$subsection, this.$module.firstChild)
+    }
```
[the remark this answers is not in the document]
Comment (andysellick) [reply app/assets/javascripts/browse-columns.js:46]: Not quite sure what you mean, can you clarify?

**Comment by andysellick** [comment]
@kevindew massive thanks for the review. Yes, I appreciate this is a huge change that's difficult to follow, sorry about that. I'm quite concerned that the animations didn't work for you as expected, still seem to be working for me, can we catch up when you're free to discuss?

File: app/assets/javascripts/browse-columns.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply app/assets/javascripts/browse-columns.js:46]: Yup sure. The previous code had both an if and an else, but this code only has an if:

https://github.com/alphagov/collections/blob/40cdc6b4b03e0267b7fac972cb925bed231fb675/app/assets/javascripts/browse-columns.js#L24-L29

**Comment by kevindew** [comment]
> @kevindew massive thanks for the review. Yes, I appreciate this is a huge change that's difficult to follow, sorry about that. I'm quite concerned that the animations didn't work for you as expected, still seem to be working for me, can we catch up when you're free to discuss?

I've done some digging and it turns out I still had an earlier branch of your code on my machine - sorry for any panic caused there. After some git resetting the problem is fixed: 

Dec-01-2021 10-28-09

File: app/assets/javascripts/browse-columns.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/assets/javascripts/browse-columns.js:46]: Ah thanks, got it now. One of the changes I've made is that the three columns are now no longer shown or hidden - they're always there but depending on the parent class on the overall browse they might not be visible. So I think to answer your question, that's correct, the code no longer needs the equivalent of `this.$subsection.show()` if this element exists.

[the remark this answers is not in the document]
Comment (kevindew) [reply app/assets/javascripts/browse-columns.js:46]: Super 👍

[the remark this answers is not in the document]
Comment (andysellick) [reply app/assets/javascripts/browse-columns.js]: Actually now I come to think about it this whole function is a bit redundant as we could just put the empty elements into the view. I've done this change in a new commit if you can take a look @kevindew?

**Review by kevindew — APPROVED** [review]
Great stuff - nice to see the code removal.

**Pull request merged**